engine/verification/assigner/blockconsumer/reader.go

```diff
@@ -0,0 +1,65 @@
+package blockconsumer
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// FinalizedBlockReader provides an abstraction for consumers to read blocks as job.
+type FinalizedBlockReader struct {
+	state  protocol.State
+	blocks storage.Blocks
+}
+
+// newFinalizedBlockReader creates and returns a FinalizedBlockReader.
+func newFinalizedBlockReader(state protocol.State, blocks storage.Blocks) *FinalizedBlockReader {
+	return &FinalizedBlockReader{
+		state:  state,
+		blocks: blocks,
+	}
+}
+
+// AtIndex returns the block job at the given index.
+// The block job at an index is just the finalized block at that index (i.e., height).
+func (r FinalizedBlockReader) AtIndex(index int64) (module.Job, error) {
+	block, err := r.blockByHeight(uint64(index))
```

Comment:
⚠️ during the development, we realized a mismatch between data types of `consumer` and `reader` on indexing the blocks. The reader operates on `uint64` to comply with the protocol state. However `consumer` operates on `int64`. This can cause some overflow on the height of requested blocks at higher heights, and **I am addressing this in the immediate subsequent PR**. Excluding that fix in this PR was to separate the domain of concerns.